<!DOCTYPE html>
<html>

  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>利用PLINK进行GWAS分析（二）</title>
  <meta name="description" content="PLINK软件输入文件的常见格式类型：&lt;/br&gt;1，一般格式：PED/MAP&lt;/br&gt;2，转置格式：TPED/TFAM&lt;/br&gt;3，二进制格式：BED/BIM/FAM&lt;/br&gt;">

  <link rel="stylesheet" href="/css/main.css">
  <link rel="canonical" href="http://zhangyy.site/notes/2015/04/04/GWAS2file.html">
  <link rel="alternate" type="application/rss+xml" title="geneticswithme" href="http://zhangyy.site/feed.xml">
<script type="text/javascript"
  src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>


</head>


  <body>
    

    <header class="site-header">

  <div class="wrapper">

    <a class="site-title" href="/">geneticswithme</a>

    <nav class="site-nav">
      <a href="#" class="menu-icon">
        <svg viewBox="0 0 18 15">
          <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
          <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
          <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
        </svg>
      </a>

      <div class="trigger">
        
          
          <a class="page-link" href="/about/">About</a>
          
        
          
        
          
        
          
        
      </div>
    </nav>

  </div>

</header>


    <div class="page-content">
      <div class="wrapper">
        <article class="post" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title" itemprop="name headline">利用PLINK进行GWAS分析（二）</h1>
    <p class="post-meta"><time datetime="2015-04-04T00:00:00-04:00" itemprop="datePublished">Apr 4, 2015</time></p>
  </header>

  <div class="post-content" itemprop="articleBody">
    <p>PLINK软件输入文件的常见格式类型：&lt;/br&gt;
1，一般格式：PED/MAP&lt;/br&gt;
2，转置格式：TPED/TFAM&lt;/br&gt;
3，二进制格式：BED/BIM/FAM&lt;/br&gt;</p>

<p>几种格式之间可以相互转换。推荐使用BED/BIM/FAM这种格式，读取速度快。BED文件包含SNP数据，是二进制格式，不能由Notepad++等文本编辑器打开。BIM文件包括SNP位置信息，FAM文件包括家系表型信息，这两种文件都是文本格式。
PED文件格式：</p>

<figure class="highlight"><pre><code class="language-bash" data-lang="bash">column1 <span class="o">=</span> FamilyID
column2 <span class="o">=</span> IndividualID
column3 <span class="o">=</span> PaternalID
column4 <span class="o">=</span> Sex
column5 <span class="o">=</span>Phenotype <span class="o">(</span>1 <span class="o">=</span> unaffected, 2 <span class="o">=</span> affected, 0 <span class="o">=</span> missing<span class="o">)</span>
column6 + column7 <span class="o">=</span> genotype pair at SNP1
column8 + column9 <span class="o">=</span> genotype pair at SNP2
…………</code></pre></figure>

<p>例如：&lt;/br&gt;
fam1 id1 fid mid 1 1 A T G G&lt;/br&gt;
fam1 id2 fid mid 2 1 A T C G&lt;/br&gt;</p>

<p>MAP文件格式：</p>

<figure class="highlight"><pre><code class="language-bash" data-lang="bash">column1 <span class="o">=</span> Chromosome
column2 <span class="o">=</span> SNPIdentifier
column3 <span class="o">=</span> Genetic Distance <span class="k">in </span>morgans<span class="o">(</span>0, <span class="k">if </span>missing<span class="o">)</span>
column4 <span class="o">=</span> Physical base-pair position <span class="k">in </span>bp units
<span class="c"># column3 and column4 are not required for basic association testing.</span></code></pre></figure>

<p>MAP文件中染色体编号是根据人类染色体设计的：&lt;/br&gt;
1-22：常染色体&lt;/br&gt;
23：X染色体&lt;/br&gt;
24: Y染色体&lt;/br&gt;
25：XY染色体拟常染色体区&lt;/br&gt;
26：线粒体&lt;/br&gt;
PLINK1.07中<code class="highlighter-rouge">--sheep</code>,<code class="highlighter-rouge">--cow</code>,<code class="highlighter-rouge">--horse</code>,<code class="highlighter-rouge">--mouse</code>,<code class="highlighter-rouge">--dog</code>，可以根据这几种的动物基因组设置染色体编号。&lt;/br&gt;
读取绵羊SNP数据可以用<code class="highlighter-rouge">plink --file test --sheep</code></p>

<p>PLINK1.09中<code class="highlighter-rouge">--chr-set 26</code>，设定1-26号染色体为常染色体，27号为X染色体，28号为Y染色体。</p>

<p>输入文件格式彼此转换的方法：&lt;/br&gt;
1、PED/MAP 转换为TPED/TFAM格式</p>

<figure class="highlight"><pre><code class="language-bash" data-lang="bash">plink --ped test.ped --map test.map  --recode --transpose --out test1
或者
plink --file <span class="nb">test</span> --recode --transpose --out test1
<span class="c">#生成test1.tped和test1.tfam文件</span></code></pre></figure>

<p>2、TPED/TFAM转化为PED/MAP文件</p>

<figure class="highlight"><pre><code class="language-bash" data-lang="bash">plink --tped test1.tped --tfam test1.tfam  --recode --out test2
或者
plink --tfile test1 --recode --out test2
<span class="c">#生成test2.ped和test2.map文件</span></code></pre></figure>

<p>3、生成二进制格式输入文件</p>

<figure class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="c">#PED/MAP转为二进制格式</span>
plink --file <span class="nb">test</span> --make-bed --out test3
<span class="c">#TFAM/TPED转为二进制格式</span>
plink --tfile test1 --make-bed --out test3
<span class="c">#生成test3.bed，test3.bim和test3.fam文件</span></code></pre></figure>

<p>4、二进制格式转为PED/MAP或TPED/TFAM</p>

<figure class="highlight"><pre><code class="language-bash" data-lang="bash"><span class="c">#用bfile来读取test3.bed，test3.bim和test3.fam文件</span>
plink --bfile test3  --recode --transpose --out test4
<span class="c">#生成test4.tped和test4.tfam</span>

plink --bfile test3 --recode --out test5
<span class="c">#生成test5.ped和test5.fam</span></code></pre></figure>

<p>其他格式转换命令：&lt;/br&gt;
<code class="highlighter-rouge">--recodeAD</code>，SNP编码成加性显性模式，以0、1、2编码SNP，NA为缺失值；&lt;/br&gt;<code class="highlighter-rouge">--recode12</code>，SNP编码为数字1或2，缺失值为0.</p>


  </div>
  <div id="disqus_thread"></div>
<script>
var disqus_config = function () {
this.page.url = page.url;
this.page.identifier = page.id;
};
(function() { // DON'T EDIT BELOW THIS LINE
var d = document, s = d.createElement('script');
s.src = '//EXAMPLE.disqus.com/embed.js';
s.setAttribute('data-timestamp', +new Date());
(d.head || d.body).appendChild(s);
})();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>

</article>

      </div>
    </div>

    <footer class="site-footer">

  <div class="wrapper">

    <h2 class="footer-heading">geneticswithme</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li>geneticswithme</li>
          <li><a href="mailto:yyzhang@cau.edu.cn">yyzhang@cau.edu.cn</a></li>
        </ul>
      </div>

      <div class="footer-col footer-col-2">
        <ul class="social-media-list">
          
          <li>
            <a href="https://github.com/geneticswithme"><span class="icon icon--github"><svg viewBox="0 0 16 16"><path fill="#828282" d="M7.999,0.431c-4.285,0-7.76,3.474-7.76,7.761 c0,3.428,2.223,6.337,5.307,7.363c0.388,0.071,0.53-0.168,0.53-0.374c0-0.184-0.007-0.672-0.01-1.32 c-2.159,0.469-2.614-1.04-2.614-1.04c-0.353-0.896-0.862-1.135-0.862-1.135c-0.705-0.481,0.053-0.472,0.053-0.472 c0.779,0.055,1.189,0.8,1.189,0.8c0.692,1.186,1.816,0.843,2.258,0.645c0.071-0.502,0.271-0.843,0.493-1.037 C4.86,11.425,3.049,10.76,3.049,7.786c0-0.847,0.302-1.54,0.799-2.082C3.768,5.507,3.501,4.718,3.924,3.65 c0,0,0.652-0.209,2.134,0.796C6.677,4.273,7.34,4.187,8,4.184c0.659,0.003,1.323,0.089,1.943,0.261 c1.482-1.004,2.132-0.796,2.132-0.796c0.423,1.068,0.157,1.857,0.077,2.054c0.497,0.542,0.798,1.235,0.798,2.082 c0,2.981-1.814,3.637-3.543,3.829c0.279,0.24,0.527,0.713,0.527,1.437c0,1.037-0.01,1.874-0.01,2.129 c0,0.208,0.14,0.449,0.534,0.373c3.081-1.028,5.302-3.935,5.302-7.362C15.76,3.906,12.285,0.431,7.999,0.431z"/></svg>
</span><span class="username">geneticswithme</span></a>

          </li>
          

          
          <li>
            <a href="https://twitter.com/quz_44"><span class="icon icon--twitter"><svg viewBox="0 0 16 16"><path fill="#828282" d="M15.969,3.058c-0.586,0.26-1.217,0.436-1.878,0.515c0.675-0.405,1.194-1.045,1.438-1.809c-0.632,0.375-1.332,0.647-2.076,0.793c-0.596-0.636-1.446-1.033-2.387-1.033c-1.806,0-3.27,1.464-3.27,3.27 c0,0.256,0.029,0.506,0.085,0.745C5.163,5.404,2.753,4.102,1.14,2.124C0.859,2.607,0.698,3.168,0.698,3.767 c0,1.134,0.577,2.135,1.455,2.722C1.616,6.472,1.112,6.325,0.671,6.08c0,0.014,0,0.027,0,0.041c0,1.584,1.127,2.906,2.623,3.206 C3.02,9.402,2.731,9.442,2.433,9.442c-0.211,0-0.416-0.021-0.615-0.059c0.416,1.299,1.624,2.245,3.055,2.271 c-1.119,0.877-2.529,1.4-4.061,1.4c-0.264,0-0.524-0.015-0.78-0.046c1.447,0.928,3.166,1.469,5.013,1.469 c6.015,0,9.304-4.983,9.304-9.304c0-0.142-0.003-0.283-0.009-0.423C14.976,4.29,15.531,3.714,15.969,3.058z"/></svg>
</span><span class="username">quz_44</span></a>

          </li>
          
        </ul>
      </div>

      <div class="footer-col footer-col-3">
        <p>Yuanyuan Zhang, a ph.D student struggling on Animal Genetics.
</p>
      </div>
    </div>

  </div>

</footer>


  </body>

</html>
