<!DOCTYPE html>
<html>

  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>利用模拟NGS数据研究kmers的分布</title>
  <meta name="description" content="wgsim, a tool for simulating NGS reads from reference sequences.">

  <link rel="stylesheet" href="/css/main.css">
  <link rel="canonical" href="http://zhangyy.site/notes/2017/08/08/wgsim_kmers.html">
  <link rel="alternate" type="application/rss+xml" title="geneticswithme" href="http://zhangyy.site/feed.xml">
<script type="text/javascript"
  src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>


</head>


  <body>
    

    <header class="site-header">

  <div class="wrapper">

    <a class="site-title" href="/">geneticswithme</a>

    <nav class="site-nav">
      <a href="#" class="menu-icon">
        <svg viewBox="0 0 18 15">
          <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
          <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
          <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
        </svg>
      </a>

      <div class="trigger">
        
          
          <a class="page-link" href="/about/">About</a>
          
        
          
        
          
        
          
        
      </div>
    </nav>

  </div>

</header>


    <div class="page-content">
      <div class="wrapper">
        <article class="post" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title" itemprop="name headline">利用模拟NGS数据研究kmers的分布</h1>
    <p class="post-meta"><time datetime="2017-08-08T00:00:00-04:00" itemprop="datePublished">Aug 8, 2017</time></p>
  </header>

  <div class="post-content" itemprop="articleBody">
    <p>wgsim, a tool for simulating NGS reads from reference sequences.</p>

<div class="highlighter-rouge"><pre class="highlight"><code>
Usage:   wgsim [options] &lt;in.ref.fa&gt; &lt;out.read1.fq&gt; &lt;out.read2.fq&gt;

Options: -e FLOAT      base error rate [0.020]
         -d INT        outer distance between the two ends [500]
         -s INT        standard deviation [50]
         -N INT        number of read pairs [1000000]
         -1 INT        length of the first read [70]
         -2 INT        length of the second read [70]
         -r FLOAT      rate of mutations [0.0010]
         -R FLOAT      fraction of indels [0.15]
         -X FLOAT      probability an indel is extended [0.30]
         -S INT        seed for random generator [0, use the current time]
         -A FLOAT      discard if the fraction of ambiguous bases higher than FLOAT [0.05]
         -h            haplotype mode
</code></pre>
</div>
<p>以绵羊(Ovis aries)的24号染色体序列为例，模拟20×的 Illumina PE reads， read length=100，假设没有base error。
number_of_reads（N）= 20× length_of_ref/(read_length × 2) =4.2e6</p>

<div class="highlighter-rouge"><pre class="highlight"><code>wgsim -e 0 -1 100 -2 100 -S 44 -N 4200000 oar_chr24.fa test_F.fq test_R.fq

</code></pre>
</div>

<p>使用dsk计算K-mers的coverage distribution，取K-mer size=31，</p>

<div class="highlighter-rouge"><pre class="highlight"><code># k-mers distribution of reference chromosome, oar_chr24
dsk -file oar_chr24.fa -kmer-size 31 -histo-max 200 -out oar_chr24
# output file: oar_chr24.h5

h5dump -y -d histogram/histogram oar_chr24.h5 | grep "^\ *[0-9]" |tr -d ", " | paste - - &gt;oar_chr24_kmerSize31_coverage.txt

# k-mers distribution of simulated reads
dsk -file test_F.fq,test_R.fq -kmer-size 42 -histo-max 200 -out test
# output file: test.h5

h5dump -y -d histogram/histogram test.h5 | grep "^\ *[0-9]" |tr -d ", " | paste - - &gt;test_kmerSize31_coverage.txt

</code></pre>
</div>

<p>接下来分别看kmer size=31时，chr24参考序列和模拟的reads的31-mers coverage distribution</p>

<h1 id="chr2431-mers">1) chr24参考序列的31-mers分布情况</h1>

<table>
  <thead>
    <tr>
      <th>coverage</th>
      <th>counts</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>1</td>
      <td>38458117</td>
    </tr>
    <tr>
      <td>2</td>
      <td>441456</td>
    </tr>
    <tr>
      <td>3</td>
      <td>85126</td>
    </tr>
    <tr>
      <td>4</td>
      <td>37148</td>
    </tr>
    <tr>
      <td>5</td>
      <td>21574</td>
    </tr>
    <tr>
      <td>6</td>
      <td>13682</td>
    </tr>
    <tr>
      <td>7</td>
      <td>9055</td>
    </tr>
    <tr>
      <td>8</td>
      <td>6752</td>
    </tr>
    <tr>
      <td>9</td>
      <td>5197</td>
    </tr>
    <tr>
      <td>10</td>
      <td>3896</td>
    </tr>
  </tbody>
</table>

<p>也就是说38458117个31-mers仅出现一次，441456个31-mers出现两次，85126个31-mers出现三次，多次出现的kmer可能是由于基因组上的重复序列。</p>

<h1 id="simulation-ngs-reads-2031-mers">2) simulation NGS reads （20×）的31-mers分布情况</h1>

<p>kmers coverage distribution如下图所示：
<img src="/images/simulation_kmers31.png" alt="kmers_simulation" /></p>

<p>模拟的NGS数据的31-kmer的分布，peak出现在coverage=14的位置。因为模拟的reads长度为100nt，则每一条read最多可能有100-31+1=70个不同的31-kmers，模拟的是20×的NGS数据量，绝大多数的31-kmers在参考序列中是unique的，所以 simulation NGS reads的31-mers distribution的peak出现就在X轴（coverage）14（70×20/100=14）的位置。</p>

<p>如果模拟的NGS reads不是 error free，31-mers分布可能会出现偏移。</p>

  </div>
 <div id="disqus_thread"></div>
<script>

/**
*  RECOMMENDED CONFIGURATION VARIABLES: EDIT AND UNCOMMENT THE SECTION BELOW TO INSERT DYNAMIC VALUES FROM YOUR PLATFORM OR CMS.
*  LEARN WHY DEFINING THESE VARIABLES IS IMPORTANT: https://disqus.com/admin/universalcode/#configuration-variables*/
/*
var disqus_config = function () {
this.page.url = PAGE_URL;  // Replace PAGE_URL with your page's canonical URL variable
this.page.identifier = PAGE_IDENTIFIER; // Replace PAGE_IDENTIFIER with your page's unique identifier variable
};
*/
(function() { // DON'T EDIT BELOW THIS LINE
var d = document, s = d.createElement('script');
s.src = 'https://zhangyy-site.disqus.com/embed.js';
s.setAttribute('data-timestamp', +new Date());
(d.head || d.body).appendChild(s);
})();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
                            

</article>

      </div>
    </div>

    <footer class="site-footer">

  <div class="wrapper">

    <h2 class="footer-heading">geneticswithme</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li>geneticswithme</li>
          <li><a href="mailto:yyzhang@cau.edu.cn">yyzhang@cau.edu.cn</a></li>
        </ul>
      </div>

      <div class="footer-col footer-col-2">
        <ul class="social-media-list">
          
          <li>
            <a href="https://github.com/geneticswithme"><span class="icon icon--github"><svg viewBox="0 0 16 16"><path fill="#828282" d="M7.999,0.431c-4.285,0-7.76,3.474-7.76,7.761 c0,3.428,2.223,6.337,5.307,7.363c0.388,0.071,0.53-0.168,0.53-0.374c0-0.184-0.007-0.672-0.01-1.32 c-2.159,0.469-2.614-1.04-2.614-1.04c-0.353-0.896-0.862-1.135-0.862-1.135c-0.705-0.481,0.053-0.472,0.053-0.472 c0.779,0.055,1.189,0.8,1.189,0.8c0.692,1.186,1.816,0.843,2.258,0.645c0.071-0.502,0.271-0.843,0.493-1.037 C4.86,11.425,3.049,10.76,3.049,7.786c0-0.847,0.302-1.54,0.799-2.082C3.768,5.507,3.501,4.718,3.924,3.65 c0,0,0.652-0.209,2.134,0.796C6.677,4.273,7.34,4.187,8,4.184c0.659,0.003,1.323,0.089,1.943,0.261 c1.482-1.004,2.132-0.796,2.132-0.796c0.423,1.068,0.157,1.857,0.077,2.054c0.497,0.542,0.798,1.235,0.798,2.082 c0,2.981-1.814,3.637-3.543,3.829c0.279,0.24,0.527,0.713,0.527,1.437c0,1.037-0.01,1.874-0.01,2.129 c0,0.208,0.14,0.449,0.534,0.373c3.081-1.028,5.302-3.935,5.302-7.362C15.76,3.906,12.285,0.431,7.999,0.431z"/></svg>
</span><span class="username">geneticswithme</span></a>

          </li>
          

          
          <li>
            <a href="https://twitter.com/quz_44"><span class="icon icon--twitter"><svg viewBox="0 0 16 16"><path fill="#828282" d="M15.969,3.058c-0.586,0.26-1.217,0.436-1.878,0.515c0.675-0.405,1.194-1.045,1.438-1.809c-0.632,0.375-1.332,0.647-2.076,0.793c-0.596-0.636-1.446-1.033-2.387-1.033c-1.806,0-3.27,1.464-3.27,3.27 c0,0.256,0.029,0.506,0.085,0.745C5.163,5.404,2.753,4.102,1.14,2.124C0.859,2.607,0.698,3.168,0.698,3.767 c0,1.134,0.577,2.135,1.455,2.722C1.616,6.472,1.112,6.325,0.671,6.08c0,0.014,0,0.027,0,0.041c0,1.584,1.127,2.906,2.623,3.206 C3.02,9.402,2.731,9.442,2.433,9.442c-0.211,0-0.416-0.021-0.615-0.059c0.416,1.299,1.624,2.245,3.055,2.271 c-1.119,0.877-2.529,1.4-4.061,1.4c-0.264,0-0.524-0.015-0.78-0.046c1.447,0.928,3.166,1.469,5.013,1.469 c6.015,0,9.304-4.983,9.304-9.304c0-0.142-0.003-0.283-0.009-0.423C14.976,4.29,15.531,3.714,15.969,3.058z"/></svg>
</span><span class="username">quz_44</span></a>

          </li>
          
        </ul>
      </div>

      <div class="footer-col footer-col-3">
        <p>Yuanyuan Zhang, a ph.D student struggling on Animal Genetics.
</p>
      </div>
    </div>

  </div>

</footer>

    <script id="dsq-count-scr" src="//zhangyy-site.disqus.com/count.js" async></script>

  </body>

</html>
