<!DOCTYPE html>
<html>

  <head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>samtools简明操作指南</title>
  <meta name="description" content="#samtools">

  <link rel="stylesheet" href="/css/main.css">
  <link rel="canonical" href="http://zhangyy.site/notes/2016/08/14/samtools.html">
  <link rel="alternate" type="application/rss+xml" title="geneticswithme" href="http://zhangyy.site/feed.xml">
<script type="text/javascript"
  src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>


</head>


  <body>
    

    <header class="site-header">

  <div class="wrapper">

    <a class="site-title" href="/">geneticswithme</a>

    <nav class="site-nav">
      <a href="#" class="menu-icon">
        <svg viewBox="0 0 18 15">
          <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
          <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
          <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
        </svg>
      </a>

      <div class="trigger">
        
          
          <a class="page-link" href="/about/">About</a>
          
        
          
        
          
        
          
        
      </div>
    </nav>

  </div>

</header>


    <div class="page-content">
      <div class="wrapper">
        <article class="post" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title" itemprop="name headline">samtools简明操作指南</h1>
    <p class="post-meta"><time datetime="2016-08-14T00:00:00-04:00" itemprop="datePublished">Aug 14, 2016</time></p>
  </header>

  <div class="post-content" itemprop="articleBody">
    <p>#samtools</p>

<ul>
  <li>
    <p>计算覆盖度</p>
  </li>
  <li>
    <p>计算比对率</p>
  </li>
</ul>

<h2 id="reads">1 统计  参考序列的长度，比对上的reads数目</h2>

<div class="highlighter-rouge"><pre class="highlight"><code>
samtools idxstats in.sam &gt;in.stat

cat in.stat

#    chr1 16616 8633270 0

#    chr2 275406953 3220386 0

#    .....

samtools view -c in.sam #直接输出 number of alignment 

    #44385909

</code></pre>
</div>

<p>the output is TAB-delimited with each line consisting of reference sequence name, sequece length, #mapped reads and #unmapped reads. it is written to stdout.</p>

<p><em>in.stat</em> 中第二列为每条染色体的长度，第三列为比对上的reads数，即alignment数目，因为某些reads会比对至参考序列中的多个位置，所以alignment数目比实际上的reads数目多。</p>

<p>如果用&lt;font color=red&gt;bowtie&lt;/font&gt;进行mapping，align_summary.txt中的比对记录与samtools idxstats的结果会不一致。</p>

<p><code class="highlighter-rouge"> samtools view -c in.bam</code>得到的reads数目同idxstats的结果一致，都是alignments 数目，比实际reads数目多（因为部分reads有多个alignments）</p>

<p>##2 samtools中的flag</p>

<table>
  <tbody>
    <tr>
      <td>Flag</td>
      <td>Chr</td>
      <td>Description</td>
    </tr>
  </tbody>
</table>

<p>|————|————-|</p>

<table>
  <tbody>
    <tr>
      <td>0x0001</td>
      <td>p</td>
      <td>the read is paired in sequencing</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x0002</td>
      <td>P</td>
      <td>the read is mapped in a proper pair</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x0004</td>
      <td>u</td>
      <td>the query sequence itself is unmapped</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x0008</td>
      <td>U</td>
      <td>the mate is unmapped</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x0010</td>
      <td>R</td>
      <td>strand of the mate</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x0040</td>
      <td>1</td>
      <td>the read is the first read in a pair</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x0080</td>
      <td>2</td>
      <td>the read is the second read in a pair</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x0100</td>
      <td>s</td>
      <td>the alignment is not primary</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x0200</td>
      <td>f</td>
      <td>the read fails platform/vendor quality checks</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x0400</td>
      <td>d</td>
      <td>the read is either a PCR or an optical duplicate</td>
    </tr>
  </tbody>
</table>

<p>新的document</p>

<table>
  <tbody>
    <tr>
      <td>flag</td>
      <td>CHR</td>
      <td>Description</td>
    </tr>
  </tbody>
</table>

<p>—|—|—</p>

<table>
  <tbody>
    <tr>
      <td>0x1</td>
      <td>PAIRED</td>
      <td>paired-end (or multiple-segment) sequencing technology</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x2</td>
      <td>PROPER_PAIR</td>
      <td>each segment properly aligned according to the aligner</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x4</td>
      <td>UNMAP</td>
      <td>segment unmapped</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x8</td>
      <td>MUNMAP</td>
      <td>next segment in the template unmapped</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x10</td>
      <td>REVERSE</td>
      <td>SEQ is reverse complemented</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x20</td>
      <td>MREVERSE</td>
      <td>SEQ of the next segment in the template is reverse complemented</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x40</td>
      <td>READ1</td>
      <td>the first segment in the template</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x80</td>
      <td>READ2</td>
      <td>the last segment in the template</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x100</td>
      <td>SECONDARY</td>
      <td>secondary alignment</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x200</td>
      <td>QCFAIL</td>
      <td>not passing quality controls</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x400</td>
      <td>DUP</td>
      <td>PCR or optical duplicate</td>
    </tr>
  </tbody>
</table>

<table>
  <tbody>
    <tr>
      <td>0x800</td>
      <td>SUPPLEMENTARY</td>
      <td>supplementary alignment</td>
    </tr>
  </tbody>
</table>

<p>###2.1 samtools view -f 和 -F</p>

<p>samtools中的view工具使用&lt;font color=red&gt;-f INT&lt;/font&gt;来依照INT与标记致保留reads，使用&lt;font color=red&gt;-F INT&lt;/font&gt;来跳过reads。当然我们还可使用它们来过滤其它的信息。比如我们想知道有多少paired end reads有mate并且都有map时，可以使用-f 1 -F 12来过滤。</p>

<div class="highlighter-rouge"><pre class="highlight"><code>

samtools view -c -f 1 -F 12 test.bam
</code></pre>
</div>

<p>其中-f 1指定只包含那些paired end reads，-F 12是不包含那些unmapped(flag 0x0004)以及mate是unmapped(flag 0x0008)。0x0004 + 0x0008 = 12.</p>

<p>###2.2 samtools flagstat</p>

<div class="highlighter-rouge"><pre class="highlight"><code>
samtools flagstat test.bam

 44385909 + 0 in total (QC-passed reads + QC-failed reads) #总共的alignment read数（QC-passed + QC-failed）

 3877817 + 0 secondary #

 0 + 0 supplementary

 0 + 0 duplicates

 44385909 + 0 mapped (100.00%:nan%) #

 40508090 + 0 paired in sequencing

 20371796 + 0 read1

 20136294 + 0 read2

 36191000 + 0 properly paired (89.34%:nan%)

 38130076 + 0 with itself and mate mapped

 2378014 + 0 singletons (5.87%:nan%)

 1057880 + 0 with mate mapped to a different chr

 146040 + 0 with mate mapped to a different chr (mapQ&gt;=5)

</code></pre>
</div>

<h3 id="calculate-depth">calculate depth</h3>

<div class="highlighter-rouge"><pre class="highlight"><code>

samtools depth -a *bamfile*  |  awk '{sum+=$3} END { print "Average = ",sum/NR}'


</code></pre>
</div>

<p>#vcftools、bcftools</p>

<ul>
  <li>提取高质量位点</li>
</ul>

<div class="highlighter-rouge"><pre class="highlight"><code>
samtools mpileup -g -f genomes/NC_008253.fna alignments/sim_reads_aligned.sorted.bam &gt; variants/sim_variants.bcf
bcftools call -c -v variants/sim_variants.bcf &gt; variants/sim_variants.vcf
</code></pre>
</div>


  </div>
  <div id="disqus_thread"></div>
<script>
var disqus_config = function () {
this.page.url = page.url;
this.page.identifier = page.id;
};
(function() { // DON'T EDIT BELOW THIS LINE
var d = document, s = d.createElement('script');
s.src = '//EXAMPLE.disqus.com/embed.js';
s.setAttribute('data-timestamp', +new Date());
(d.head || d.body).appendChild(s);
})();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>

</article>

      </div>
    </div>

    <footer class="site-footer">

  <div class="wrapper">

    <h2 class="footer-heading">geneticswithme</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li>geneticswithme</li>
          <li><a href="mailto:yyzhang@cau.edu.cn">yyzhang@cau.edu.cn</a></li>
        </ul>
      </div>

      <div class="footer-col footer-col-2">
        <ul class="social-media-list">
          
          <li>
            <a href="https://github.com/geneticswithme"><span class="icon icon--github"><svg viewBox="0 0 16 16"><path fill="#828282" d="M7.999,0.431c-4.285,0-7.76,3.474-7.76,7.761 c0,3.428,2.223,6.337,5.307,7.363c0.388,0.071,0.53-0.168,0.53-0.374c0-0.184-0.007-0.672-0.01-1.32 c-2.159,0.469-2.614-1.04-2.614-1.04c-0.353-0.896-0.862-1.135-0.862-1.135c-0.705-0.481,0.053-0.472,0.053-0.472 c0.779,0.055,1.189,0.8,1.189,0.8c0.692,1.186,1.816,0.843,2.258,0.645c0.071-0.502,0.271-0.843,0.493-1.037 C4.86,11.425,3.049,10.76,3.049,7.786c0-0.847,0.302-1.54,0.799-2.082C3.768,5.507,3.501,4.718,3.924,3.65 c0,0,0.652-0.209,2.134,0.796C6.677,4.273,7.34,4.187,8,4.184c0.659,0.003,1.323,0.089,1.943,0.261 c1.482-1.004,2.132-0.796,2.132-0.796c0.423,1.068,0.157,1.857,0.077,2.054c0.497,0.542,0.798,1.235,0.798,2.082 c0,2.981-1.814,3.637-3.543,3.829c0.279,0.24,0.527,0.713,0.527,1.437c0,1.037-0.01,1.874-0.01,2.129 c0,0.208,0.14,0.449,0.534,0.373c3.081-1.028,5.302-3.935,5.302-7.362C15.76,3.906,12.285,0.431,7.999,0.431z"/></svg>
</span><span class="username">geneticswithme</span></a>

          </li>
          

          
          <li>
            <a href="https://twitter.com/quz_44"><span class="icon icon--twitter"><svg viewBox="0 0 16 16"><path fill="#828282" d="M15.969,3.058c-0.586,0.26-1.217,0.436-1.878,0.515c0.675-0.405,1.194-1.045,1.438-1.809c-0.632,0.375-1.332,0.647-2.076,0.793c-0.596-0.636-1.446-1.033-2.387-1.033c-1.806,0-3.27,1.464-3.27,3.27 c0,0.256,0.029,0.506,0.085,0.745C5.163,5.404,2.753,4.102,1.14,2.124C0.859,2.607,0.698,3.168,0.698,3.767 c0,1.134,0.577,2.135,1.455,2.722C1.616,6.472,1.112,6.325,0.671,6.08c0,0.014,0,0.027,0,0.041c0,1.584,1.127,2.906,2.623,3.206 C3.02,9.402,2.731,9.442,2.433,9.442c-0.211,0-0.416-0.021-0.615-0.059c0.416,1.299,1.624,2.245,3.055,2.271 c-1.119,0.877-2.529,1.4-4.061,1.4c-0.264,0-0.524-0.015-0.78-0.046c1.447,0.928,3.166,1.469,5.013,1.469 c6.015,0,9.304-4.983,9.304-9.304c0-0.142-0.003-0.283-0.009-0.423C14.976,4.29,15.531,3.714,15.969,3.058z"/></svg>
</span><span class="username">quz_44</span></a>

          </li>
          
        </ul>
      </div>

      <div class="footer-col footer-col-3">
        <p>Yuanyuan Zhang, a ph.D student struggling on Animal Genetics.
</p>
      </div>
    </div>

  </div>

</footer>


  </body>

</html>
